真偽値でないキーワードも真偽値として解析されることがある

`atrue` など `true` または `false` の前に文字がある場合でも真偽値として解析される。

## テストコード

```go
func TestInvalidKeyword(t *testing.T) {
	got, err := parser.Analyze([]byte("atrue"))
	if !errors.Is(parser.ErrUndefinedKeyword, err) {
		t.Errorf("want: %v, got: %v", parser.ErrUndefinedKeyword, err)
	}
	if got != nil {
		t.Errorf("want: <nil>, got: %v", got)
	}
}
```

## 実行結果

```shell
mini-parser-go on git main [$!] via go v1.22.5
➜ go test ./...
--- FAIL: TestInvalidKeyword (0.00s)
    analyzer_test.go:57: want: undefined keyword, got: <nil>
    analyzer_test.go:60: want: <nil>, got: [{4 true}]
FAIL
FAIL    github.com/YO-RO/mini-parser-go/parser  0.113s
FAIL
```

--- a/parser/analyzer.go
+++ b/parser/analyzer.go
@@ -69,9 +69,13 @@ func (mt MarkToken) Value() (any, error) {
 }
 
 var (
-	ErrSyntax           error = errors.New("invalid syntax")
-	ErrUndefinedKeyword error = errors.New("undefined keyword")
+	ErrSyntax          error = errors.New("invalid syntax")
+	ErrUndefinedSymbol error = errors.New("undefined symbol")
 )
+
+func mayBeString(str string, i int) bool {
+	return str[i] == '"'
+}
 
 func extractStringAsToken(str string, startIdx int) (ValueToken, int, error) {
 	// 直前に\がない"
@@ -94,6 +98,11 @@ func extractStringAsToken(str string, startIdx int) (ValueToken, int, error) {
 		value:     value,
 	}
 	return token, endIdx, nil
+}
+
+func isNumber(str string, i int) bool {
+	matched, _ := regexp.MatchString(`\d`, str[i:i+1])
+	return matched
 }
 
 func extractNumberAsToken(str string, startIdx int) (ValueToken, int, error) {
@@ -126,13 +135,17 @@ func extractNumberAsToken(str string, startIdx int) (ValueToken, int, error) {
 	return token, startIdx + loc[1], nil
 }
 
+func mayBeBool(str string, i int) bool {
+	return str[i] == 't' || str[i] == 'f'
+}
+
 func extractBoolAsToken(str string, startIdx int) (ValueToken, int, error) {
 	// ?i: はcase insentive
 	// ?: はグループをキャプチャしない
 	re := regexp.MustCompile(`(?i:true|false)(?:[\s,:"{}\[\]]|$)`)
 	loc := re.FindStringSubmatchIndex(str[startIdx:])
 	if loc == nil {
-		return ValueToken{}, 0, ErrUndefinedKeyword
+		return ValueToken{}, 0, ErrUndefinedSymbol
 	}
 	endIdx := startIdx + loc[1]
 
@@ -150,33 +163,52 @@ func extractBoolAsToken(str string, startIdx int) (ValueToken, int, error) {
 	return token, endIdx, nil
 }
 
+func isSpace(str string, i int) bool {
+	matched, _ := regexp.MatchString(`\s`, str[i:i+1])
+	return matched
+}
+
+func skipSpaces(str string, startIdx int) int {
+	re := regexp.MustCompile(`\s+`)
+	loc := re.FindStringIndex(str[startIdx:])
+	if loc == nil {
+		return startIdx
+	}
+	return startIdx + loc[1]
+}
+
 func Analyze(d []byte) ([]Tokener, error) {
 	res := []Tokener{}
 
 	inputStr := string(d)
 	for i := 0; i < len(inputStr); i++ {
-		switch inputStr[i] {
-		case '"':
+		switch {
+		case mayBeString(inputStr, i):
 			token, endIdx, err := extractStringAsToken(inputStr, i)
 			if err != nil {
 				return nil, err
 			}
 			res = append(res, token)
 			i = endIdx - 1
-		case '1', '2', '3', '4', '5', '6', '7', '8', '9', '0':
+		case isNumber(inputStr, i):
 			token, endIdx, err := extractNumberAsToken(inputStr, i)
 			if err != nil {
 				return nil, err
 			}
 			res = append(res, token)
 			i = endIdx - 1
-		case 't', 'f':
+		case mayBeBool(inputStr, i):
 			token, endIdx, err := extractBoolAsToken(inputStr, i)
 			if err != nil {
 				return nil, err
 			}
 			res = append(res, token)
 			i = endIdx - 1
+		case isSpace(inputStr, i):
+			endIdx := skipSpaces(inputStr, i)
+			i = endIdx - 1
+		default:
+			return nil, ErrUndefinedSymbol
 		}
 	}
 

--- a/parser/analyzer_test.go
+++ b/parser/analyzer_test.go
@@ -71,6 +71,18 @@ func TestBoolAnalyzer(t *testing.T) {
 			parser.ErrUndefinedKeyword,
 			nil,
 		},
+		{
+			"定義されていないキーワード[atrue]",
+			"atrue",
+			parser.ErrUndefinedKeyword,
+			nil,
+		},
+		{
+			"定義されていないキーワード[afalse]",
+			"afalse",
+			parser.ErrUndefinedKeyword,
+			nil,
+		},
 	}
 	runTestCases(t, tests)
 }

--- a/parser/analyzer_test.go
+++ b/parser/analyzer_test.go
@@ -68,19 +68,19 @@ func TestBoolAnalyzer(t *testing.T) {
 		{
 			"定義されていないキーワード[trueeeee]",
 			"trueeeee",
-			parser.ErrUndefinedKeyword,
+			parser.ErrUndefinedSymbol,
 			nil,
 		},
 		{
 			"定義されていないキーワード[atrue]",
 			"atrue",
-			parser.ErrUndefinedKeyword,
+			parser.ErrUndefinedSymbol,
 			nil,
 		},
 		{
 			"定義されていないキーワード[afalse]",
 			"afalse",
-			parser.ErrUndefinedKeyword,
+			parser.ErrUndefinedSymbol,
 			nil,
 		},
 	}

--- a/parser/analyzer_test.go
+++ b/parser/analyzer_test.go
@@ -83,6 +83,12 @@ func TestBoolAnalyzer(t *testing.T) {
 			parser.ErrUndefinedSymbol,
 			nil,
 		},
+		{
+			"定義されていないキーワード[true111]",
+			"true111",
+			parser.ErrUndefinedSymbol,
+			nil,
+		},
 	}
 	runTestCases(t, tests)
 }

